hist/hist/inc/THnChain.h

```diff
@@ -0,0 +1,116 @@
+// @(#)root/hist:$Id$
+// Author: Benjamin Bannier, August 2016
+
+/*************************************************************************
+ * Copyright (C) 1995-2016, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT_THnChain
+#define ROOT_THnChain
+
+#include <TObject.h>
+
+#include <string>
+#include <vector>
+
+class TAxis;
+class TH1;
+class TH2;
+class TH3;
+class THnBase;
+
+/** \class THnChain
+A class to chain together multiple histograms.
+
+This class allows to chain together any `THnBase`-derived (`THn` or `THnSparse`)
+histograms from multiple files. Operations on the axes and projections are
+supported. The intent is to allow convenient merging merging of projections
+of high-dimensional histograms.
+
+\code{.cpp}
+// `file1.root` and `file2.root` contain a `THnSparse` named `hsparse`.
+THnChain hs("hsparse");
+hs.AddFile("file1.root");
+hs.AddFile("file2.root");
+
+// Project out axis 0, integrate over other axes.
+TH1* h0 = hs.Projection(0);
+
+// Project out axis 0, integrate over other axes in their active ranges.
+hs.GetAxis(1)->SetRangeUser(0, 0.1); // select a subrange
+TH1* h0 = hs.Projection(0);
+\endcode
+*/
+
+class THnChain : public TObject
+{
+ public:
+   /// Default constructor.
+   ///
+   /// \param name name of the histogram to work on
+   THnChain(const char* name) : fName(name) {}
+
+   /// Add a new file to this chain.
+   ///
+   /// \param fileName path of the file to add
+   void AddFile(const char* fileName);
+
+   /// Get an axis from the histogram.
+   ///
+   /// \param i index of the axis to retrieve
+   ///
+   /// This function requires that a file containing the histogram was
+   /// already added with `AddFile`.
+   ///
+   /// Properties set on the axis returned by `GetAxis` are propagated to all
+   /// histograms in the chain, so it can e.g. be used to set ranges for
+   /// projections.
+   TAxis* GetAxis(Int_t i);
+
+   /// See `THnBase::Projection` for the intended behavior.
+   TH1* Projection(Int_t xDim, Option_t* option = "");
```

Comment:
I was confused here, and `SetupAxes` can be `const` w/o requiring additional internal protections. We hold some axes in the `fAxes` member var, but then copy the information of these axes to histograms from files we open while performing the projection. There is no need to protect any `fAxes` as these are not modified in `SetupAxes` (we call something like `fAxes[i]->Copy(*ax_from_file)`).

We do modify the axes of histograms we read on the fly from freshly opened files, but since the files are opened just for reading this might be safe (or at least as safe as we can make it here).

> Do you need to call SetupAxes()?

Yes, it looks like this is required to make projections actually work, e.g., the test cases added as part of https://github.com/root-mirror/roottest/pull/6 complain if `SetupAxes` is not called:

    [ RUN      ] THnChainTest.EmptySubRange
    Warning in <TROOT::Append>: Replacing existing TH1: h_proj_0 (Potential memory leak).
    /roottest/root/hist/chain/thnchain.cxx:88: Failure
    Value of: projection->GetNbinsX()
      Actual: 10
    Expected: 1
    [  FAILED  ] THnChainTest.EmptySubRange (11 ms)